runtime/lua/vim/lsp/completion.lua

```diff
@@ -0,0 +1,773 @@
+  local line_to_cursor = line:sub(1, cursor_col)
+  local clients = (buf_handles[bufnr] or {}).clients or {}
+  local word_boundary = vim.fn.match(line_to_cursor, '\\k*$')
+  local start_time = vim.uv.hrtime()
+  Context.last_request_time = start_time
+
+  local cancel_request = request(clients, bufnr, win, function(responses)
+    local end_time = vim.uv.hrtime()
+    rtt_ms = compute_new_average((end_time - start_time) * ns_to_ms)
+
+    Context.pending_requests = {}
+    Context.isIncomplete = false
+
+    local row_changed = api.nvim_win_get_cursor(win)[1] ~= cursor_row
+    local mode = api.nvim_get_mode().mode
+    if row_changed or not (mode == 'i' or mode == 'ic') then
+      return
+    end
+
+    local matches = {}
+    local server_start_boundary --- @type integer?
+    for client_id, response in pairs(responses) do
+      if response.err then
+        vim.notify_once(response.err.message, vim.log.levels.warn)
+      end
+
+      local result = response.result
+      if result then
+        Context.isIncomplete = Context.isIncomplete or result.isIncomplete
+        local client = lsp.get_client_by_id(client_id)
+        local encoding = client and client.offset_encoding or 'utf-16'
+        local client_matches
+        client_matches, server_start_boundary = M._convert_results(
+          line,
+          cursor_row - 1,
+          cursor_col,
+          client_id,
+          word_boundary,
+          nil,
+          result,
+          encoding
+        )
+        vim.list_extend(matches, client_matches)
+      end
+    end
+    local start_col = (server_start_boundary or word_boundary) + 1
+    vim.fn.complete(start_col, matches)
+  end)
+
+  table.insert(Context.pending_requests, cancel_request)
+end
+
+--- Attaches the given client to the given buffer as a completion provider.
+---
+--- @param client_id integer Client ID
+--- @param bufnr integer Buffer handle, or 0 for the current buffer
+function M.attach(client_id, bufnr)
+  bufnr = (bufnr == 0 and api.nvim_get_current_buf()) or bufnr
+
+  if not buf_handles[bufnr] then
+    buf_handles[bufnr] = { clients = {}, triggers = {} }
+
+    -- Attach to buffer events.
+    api.nvim_buf_attach(bufnr, false, {
+      on_detach = function(_, buf)
+        buf_handles[buf] = nil
+      end,
+      on_reload = function(_, buf)
+        M.attach(client_id, buf)
+      end,
+    })
+
+    -- Set up autocommands.
+    local group =
+      api.nvim_create_augroup(string.format('vim/lsp/completion-%d', bufnr), { clear = true })
+    api.nvim_create_autocmd('InsertCharPre', {
+      group = group,
+      buffer = bufnr,
+      callback = function()
+        insert_char_pre_cb(buf_handles[bufnr])
+      end,
+    })
+    api.nvim_create_autocmd('TextChangedP', {
+      group = group,
+      buffer = bufnr,
+      callback = text_changed_p_cb,
+    })
+    api.nvim_create_autocmd('TextChangedI', {
+      group = group,
+      buffer = bufnr,
+      callback = text_changed_i_cb,
+    })
```

Comment:
```suggestion
    api.nvim_create_autocmd('TextChangedP', {
      group = group,
      buffer = bufnr,
      callback = text_changed_p_cb,
    })
    api.nvim_create_autocmd('TextChangedI', {
      group = group,
      buffer = bufnr,
      callback = text_changed_i_cb,
    })
```

This is for the backspace-retrigger, right? If so, I think we can exclude this in the initial version.
It might not be needed if it's done generic for `complete()`. See https://github.com/neovim/neovim/pull/28558#issuecomment-2082451177